mlflow/pyfunc/spark_udf.py

```diff
@@ -0,0 +1,562 @@
+import collections
+import logging
+import os
+import signal
+import subprocess
+import sys
+import tempfile
+import threading
+import functools
+from typing import Union, Iterator, Tuple
+
+import numpy as np
+import pandas
+
+import mlflow
+from mlflow.environment_variables import MLFLOW_SCORING_SERVER_REQUEST_TIMEOUT
+from mlflow.exceptions import MlflowException
+from mlflow.models import Model
+from mlflow.models.flavor_backend_registry import get_flavor_backend
+from mlflow.models.model import MLMODEL_FILE_NAME
+from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE
+from mlflow.tracking.artifact_utils import _download_artifact_from_uri
+from mlflow.utils import env_manager as _EnvManager
+from mlflow.utils import find_free_port, check_port_connectivity
+from mlflow.utils.databricks_utils import is_in_databricks_runtime
+from mlflow.utils.file_utils import get_or_create_tmp_dir, get_or_create_nfs_tmp_dir
+from mlflow.utils.nfs_on_spark import get_nfs_cache_root_dir
+from mlflow.utils.requirements_utils import _warn_dependency_requirement_mismatches
+from mlflow.environment_variables import _MLFLOW_TESTING
+
+
+_logger = logging.getLogger(__name__)
+
+
+def _create_model_downloading_tmp_dir(should_use_nfs):
+    if should_use_nfs:
+        root_tmp_dir = get_or_create_nfs_tmp_dir()
+    else:
+        root_tmp_dir = get_or_create_tmp_dir()
+
+    root_model_cache_dir = os.path.join(root_tmp_dir, "models")
+    os.makedirs(root_model_cache_dir, exist_ok=True)
+
+    tmp_model_dir = tempfile.mkdtemp(dir=root_model_cache_dir)
+    # mkdtemp creates a directory with permission 0o700
+    # change it to be 0o777 to ensure it can be seen in spark UDF
+    os.chmod(tmp_model_dir, 0o777)
+    return tmp_model_dir
+
+
+_MLFLOW_SERVER_OUTPUT_TAIL_LINES_TO_KEEP = 200
+
+
+def _parse_spark_datatype(datatype: str):
+    from pyspark.sql.functions import udf
+
+    return udf(lambda x: x, returnType=datatype).returnType
+
+
+def spark_udf(spark, model_uri, result_type="double", env_manager=_EnvManager.LOCAL):
```

Comment:
Let's sync this commit https://github.com/mlflow/mlflow/commit/cee0215fb0d6982ff72fe2689084401afcd3d104 content.